server/src/repositories/Store/purchaseOrderRepository.ts

```diff
@@ -0,0 +1,177 @@
+  addItemsToPO(poId: number, items: PurchaseOrderItemInput[]): Promise<PurchaseOrder>;
+  getAttachmentsById(poId: number): Promise<string[]>;
+  addAttachmentToPO(attachment: string, poId: number): Promise<PurchaseOrder>;
+  addAttachmentsToPO(attachments: string[], poId: number): Promise<PurchaseOrder>;
+  removeAttachmentFromPO(attachment: string, poId: number): Promise<PurchaseOrder>;
+  removeItemFromPO(itemId: number, poId: number): Promise<PurchaseOrder>;
+  updateItemCountInPO(itemId: number, count: number, poId: number): Promise<PurchaseOrder>;
+  deletePOById(poId: number): Promise<void>;
+}
+
+
+export class PurchaseOrderRepository implements IPurchaseOrderRepository {
+
+  private queryBuilder
+
+  constructor(queryBuilder?: any) {
+    this.queryBuilder = queryBuilder || knex
+  }
+
+  public async getAttachmentsById(poId: number): Promise<string[]> {
+    const knexResult = await this.queryBuilder("PurchaseOrderAttachment")
+      .select(
+        "attachment"
+      )
+      .where("id", poId);
+    const result = knexResult.map((i: any) => i.attachment)
+    return result;
+  }
+
+  public async getPOItemsById(poId: number): Promise<PurchaseOrderItem[]> {
+    const knexResult = await this.queryBuilder("PurchaseOrderItem")
+      .select(
+        "id",
+        "item",
+        "count",
+        "purchaseOrder",
+      )
+      .where("purchaseOrder", poId);
+    return purchaseOrderItemsToDomain(knexResult);
+  }
+
+  public async getAllPOs(): Promise<PurchaseOrder[]> {
+    const knexResult = await this.queryBuilder("PurchaseOrder")
+      .select(
+        "id",
+        "creator",
+        "createDate",
+        "expectedDeliveryDate",
+      );
+    return purchaseOrdersToDomain(knexResult);
+  }
+
+  public async getPOById(poId: number): Promise<PurchaseOrder> {
+    const knexResult = await this.queryBuilder("PurchaseOrder")
+      .select(
+        "id",
+        "creator",
+        "createDate",
+        "expectedDeliveryDate",
+      )
+      .where("id", poId);
+    return singlePurchaseOrderToDomain(knexResult);
+  }
+
+  public async createNewPO(creatorId: number, expectedDeliveryDate: Date, items: PurchaseOrderItemInput[], attachments: string[]): Promise<PurchaseOrder> {
+    const newId = (await this.queryBuilder("PurchaseOrder")
+      .insert({
+        creator: creatorId,
+        createDate: new Date().toISOString(),
+        expectedDeliveryDate: expectedDeliveryDate,
+      }, "id"))[0];
+
+    if (items && items.length > 0) {
+      this.addItemsToPO(newId, items);
+    }
+
+    if (attachments && attachments.length > 0) {
+      this.addAttachmentsToPO(attachments, newId);
+    }
+    console.log(newId);
+    return this.getPOById(newId);
+  }
+
+  public async addItemToPO(item: InventoryItem, poId: number, count: number): Promise<PurchaseOrder> {
+    await this.queryBuilder("PurchaseOrderItem")
+      .insert({
+        item: item.id,
```

Comment:
also wondering, since the code is almost identical, if addItemToPo needs to be its own thing if we can just send a single item in an array to addItemsToPo

Reply:
Good ponint